File: persistence/src/main/java/com/yahoo/persistence/spi/PartitionState.java
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.persistence.spi;

/**
* @author thomasg
*/
public class PartitionState {
    public PartitionState(State state, String reason) {
        this.state = state;
        this.reason = reason;

        if (reason == null || state == null) {
            throw new IllegalArgumentException("State and reason must be non-null");
        }
    }

    public State getState() { return state; }
    public String getReason() { return reason; }

    State state;
    String reason;

    public enum State {
        UP,
        DOWN
    }
}

File: persistence/src/vespa/persistence/spi/partitionstate.cpp
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.

#include "partitionstate.h"
#include <vespa/vespalib/util/exceptions.h>
#include <vespa/vespalib/stllike/asciistream.h>

namespace storage::spi {

PartitionState::PartitionState()
    : _state(UP),
      _reason()
{ }

PartitionState::PartitionState(State s, vespalib::stringref reason)
    : _state(s),
      _reason(reason)
{ }

PartitionStateList::PartitionStateList(PartitionId::Type partitionCount)
    : _states(partitionCount)
{ }

PartitionStateList::~PartitionStateList() = default;

PartitionState&
PartitionStateList::operator[](PartitionId::Type index)
{
    if (index >= _states.size()) {
        vespalib::asciistream ost;
        ost << "Cannot return disk " << index << " of " << _states.size();
        throw vespalib::IllegalArgumentException(ost.str(), VESPA_STRLOC);
    }
    return _states[index];
}

}
